### PR TITLE
Fix the qr parameters object which contains a guid rather than a UUID

### PR DESCRIPTION
**Pull request by PatrickColborne**

### CODE REVIEW

**Review by CaseyODonnell92 — CHANGES_REQUESTED** [review]
I'm pretty sure this is all good. Just one small doc change. Let's only actually release and deprecate the old version when we really need to and are a bit more confident.

**Review by CaseyODonnell92 — APPROVED** [review]
Just a reminder to update release notes and pom

File: src/main/java/io/electrum/qr/api/model/QrProperties.java [hunk 1]
```diff
@@ -37,11 +44,51 @@
 
    @JsonProperty("singleUse")
    private Boolean singleUse = null;
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/QrProperties.java:46]: This field has recently been removed - feeling thankful that we waited! Please drop this guy.

**Review by CaseyODonnell92 — CHANGES_REQUESTED** [review]

File: src/main/java/io/electrum/qr/api/model/QrProperties.java [hunk 2]
```diff
@@ -202,31 +224,41 @@ public boolean equals(java.lang.Object o) {
          return false;
       }
       QrProperties qrProperties = (QrProperties) o;
-      return Objects.equals(this.uuid, qrProperties.uuid) && Objects.equals(this.value, qrProperties.value)
+      return Objects.equals(this.description, qrProperties.description)
+            && Objects.equals(this.destinationAccountId, qrProperties.destinationAccountId)
+            && Objects.equals(this.guid, qrProperties.guid) && Objects.equals(this.value, qrProperties.value)
             && Objects.equals(this.overPaymentAllowed, qrProperties.overPaymentAllowed)
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/QrProperties.java:230]: Can't seem to comment on where the field is actually defined.
Please remove `overPaymentAllowed`. The reason for this is that it intersects somewhat with how we will have to support tip amount specification later on and we aren't confident regarding what that will look like. So if we remove it now, we can always add it back at a later stage and that change won't be considered breaking.

**Review by CaseyODonnell92 — CHANGES_REQUESTED** [review]

File: src/main/java/io/electrum/qr/api/model/QrProperties.java [hunk 3]
```diff
@@ -50,14 +91,14 @@ public QrProperties uuid(String uuid) {
     *
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/QrProperties.java:91]: On the line above this, please align the Javadac with the Swagger description. Sorry I didn't catch this last time!

**Review by CaseyODonnell92 — APPROVED** [review]

File: src/main/java/io/electrum/qr/api/model/QrProperties.java [hunk 4]
```diff
@@ -20,28 +20,69 @@
...
+   private String description = null;
+
+   @JsonProperty("destinationAccountId")
+   private String destinationAccountId = null;
+
+   @JsonProperty("guid")
+   private String guid = null;
 
    @JsonProperty("value")
    private LedgerAmount value = null;
```
Comment (TimElectrum) [inline src/main/java/io/electrum/qr/api/model/QrProperties.java:34]: I think we should remove this field - there's already an `amounts` field next to the `qrProperties` in requests

**Review by TimElectrum — CHANGES_REQUESTED** [review]

File: src/main/java/io/electrum/qr/api/model/QrProperties.java [hunk 5]
```diff
@@ -20,28 +20,69 @@
...
 @ApiModel(description = "A collection of attributes that describe how a QR code is intended to be used for transacting.")
 public class QrProperties {
-   @JsonProperty("uuid")
-   private String uuid = null;
+
+   @JsonProperty("description")
+   private String description = null;
+
+   @JsonProperty("destinationAccountId")
+   private String destinationAccountId = null;
```
[the remark this answers is not in the document]
Comment (CaseyODonnell92) [reply src/main/java/io/electrum/qr/api/model/QrProperties.java:28]: The current state of the management API is such that no account IDs are present in the assign QR request. Maybe that needs to change?

[the remark this answers is not in the document]
Comment (TimElectrum) [reply src/main/java/io/electrum/qr/api/model/QrProperties.java:28]: ja, it probably does. Specifying a destination account is similar to specifying a merchant in that both identify the recipient of the payment. (In our world, at the moment, it's presumed that the owner of the QR code is meant to be who is receiving the payment.)

The info needed to request a QR is split into 2 broad groups:

1. _who_ the QR code is for i.e. the `Owner`
2. _how_ the transaction should occur i.e. the `QrProperties`

The destination account is very much part of the "who", not the "how"

[the remark this answers is not in the document]
Comment (CaseyODonnell92) [reply src/main/java/io/electrum/qr/api/model/QrProperties.java:28]: Cool, so we can easily add an account ID on Owner. We just need to add a sensible field to `CreateQrCodeRequest` to capture it at request time. And obviously it shouldn't be in `QrProperties`

[the remark this answers is not in the document]
Comment (CaseyODonnell92) [reply src/main/java/io/electrum/qr/api/model/QrProperties.java:28]: Presumably, this also means we should take the destinationAccountId out the `QrParameters` in the management API.

[the remark this answers is not in the document]
Comment (TimElectrum) [reply src/main/java/io/electrum/qr/api/model/QrProperties.java:28]: > Presumably, this also means we should take the destinationAccountId out the QrParameters in the management API.

Yip

[the remark this answers is not in the document]
Comment (TimElectrum) [reply src/main/java/io/electrum/qr/api/model/QrProperties.java:28]: > Cool, so we can easily add an account ID on Owner. We just need to add a sensible field to `CreateQrCodeRequest` to capture it at request time. And obviously it shouldn't be in `QrProperties`

I wonder if we should add `PaymentMethods` to `CreateQrCodeRequest`. The intention would be "give me a QR which can be used for this `PaymentMethod`". I think that's clear and sensible enough but it kinda goes against the design of _this_ API. If a particular `PaymentMethod` was going to be used it would only become apparent in the subsequent `PaymentRequest` so making it appear in the request for a QR code seems a little unnecessary (an example of a lack of temporal decomposition?)

But, I don't think adding `PaymentMethod` to the `CreateQrCodeRequest` actually _inhibits_ the original intention of the API (making tran details appear to everyone in the `PaymentRequest`). And with our possible intention to make the `ScanNotification` response include more transaction details, I think it would be okay to add a `PaymentMethod` to the `CreateQrCodeRequest`

File: src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java [hunk 6]
```diff
@@ -273,7 +278,8 @@ public void setCustomer(Customer customer) {
     *
     * @since 1.5.0
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java:279]:
```suggestion
    * @since 1.8.0
```

File: src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java [hunk 7]
```diff
@@ -273,7 +278,8 @@ public void setCustomer(Customer customer) {
...
     *
-    * @param qrProperties instance
+    * @param qrProperties
+    *           instance
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java:282]:
```suggestion
    * @param qrProperties
    *           the collection of attributes that describe how the requested QR code is intended to be used for 
    *           transacting.

```

File: src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java [hunk 8]
```diff
@@ -292,6 +298,36 @@ public void setQrProperties(QrProperties qrProperties) {
...
+    * @return this instance of CreateQrCodeMessage
+    **/
+   public CreateQrCodeMessage paymentMethods(@NotNull List<PaymentMethod> paymentMethods) {
+      Objects.requireNonNull(paymentMethods);
+      this.paymentMethods = paymentMethods;
+      return this;
+   }
+
+   @JsonProperty("paymentMethods")
+   @ApiModelProperty(value = "")
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java:318]:
```suggestion
   @ApiModelProperty(value = "A list of payment methods with which the QR owner will accept payment. The paymentMethods"
           + " specify the destination to which funds may be transferred when the QR code is scanned.")
```

File: src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java [hunk 9]
```diff
@@ -292,6 +298,36 @@ public void setQrProperties(QrProperties qrProperties) {
       this.qrProperties = qrProperties;
    }
 
+   /**
+    * A list of payment methods that the QR owner will accept payment with. The paymentMethods specify the destination
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java:302]:
```suggestion
    * A list of payment methods with which the QR owner will accept payment. The paymentMethods specify the destination
```

**Review by CaseyODonnell92 — CHANGES_REQUESTED** [review]
Just some really small suggestions around wording in the documentation.

File: src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java [hunk 10]
```diff
@@ -40,6 +41,7 @@
    protected Amounts amounts = null;
    protected Customer customer = null;
    protected QrProperties qrProperties = null;
+   protected List<PaymentMethod> paymentMethods = new ArrayList<>();
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java:44]: Is there some specific reason that you opted for a list here? It kind of makes sense to me that you might be able to have more than one payment method discoverable via the QR code. The only hassle I foresee is that we will just have to keep this API aligned with the QR Manager. So if we have multiple methods available here, we should _probably_ facilitate storing that info in the QR Manager.

File: src/main/java/io/electrum/qr/api/model/QrProperties.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (TimElectrum) [reply src/main/java/io/electrum/qr/api/model/QrProperties.java]: beefy conversation I don't want to resolve because I don't want it to become hidden. But happy to consider it resolved.

File: src/main/java/io/electrum/qr/api/model/QrProperties.java [hunk 11]
```diff
@@ -20,86 +18,58 @@
...
-   @JsonProperty("uuid")
-   private String uuid = null;
 
-   @JsonProperty("value")
-   private LedgerAmount value = null;
+   @JsonProperty("description")
+   private String description = null;
 
-   @JsonProperty("overPaymentAllowed")
-   private Boolean overPaymentAllowed = null;
```
Comment (TimElectrum) [inline src/main/java/io/electrum/qr/api/model/QrProperties.java:30]: I think we can keep this. This is separate to a tip and adding support for a tip doesn't negate the need for this field

File: src/main/java/io/electrum/qr/api/model/QrProperties.java [hunk 12]
```diff
@@ -20,86 +18,58 @@
...
+   public QrProperties description(String description) {
+      this.description = description;
       return this;
    }
 
    /**
-    * Get value
-    *
-    * @return value
+    * A short description for the QR code defining why it was created.
```
Comment (TimElectrum) [inline src/main/java/io/electrum/qr/api/model/QrProperties.java:40]: it might be worth adding that this is intended to be displayed to the person scanning the QR code to provide them with context for the QR

**Review by TimElectrum — CHANGES_REQUESTED** [review]
A nice project to demonstrate the importance of good tests 🙈 

The existing tests are poor in that they just use empty objects to test the broiler plate methods (there was an issue previously around "identical" instances of objects having different hash codes etc.)

If the tests used _real_ data including some of the fields you're removing then they would have started failing and it would have highlighted the breaking nature of your changes...

So there are two things we need to address (neither of which is your fault but you're working this now so...):

1. Please add some "new model" tests which test using more sensible data than the empty objects used at the moment. You can extend the class in the base API's testing package like we do with our other APIs. Please keep the "empty object" tests though - they're highlighting an important thing.
2. We need to think about how to handle these breaking changes - ideally we don't want to release a new major version or make breaking changes because that has knock-on effects elsewhere.

File: src/main/java/io/electrum/qr/api/model/QrProperties.java (hunk 2 above)
Comment (TimElectrum) [reply src/main/java/io/electrum/qr/api/model/QrProperties.java]: 1. Removing it now _is_ a breaking change and I don't think it makes sense to "remove it now because we can add it later as a non-breaking change".
2. "Over payment allowed or not" is a different question to "tip allowed or not". Think about bill payments and policy payments - sometimes an overpayment is allowed and sometimes it's not. But you'd never think about tipping!

I think `overPaymentAllowed` should stay

File: src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java (hunk 10 above)
Comment (TimElectrum) [reply src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java:44]: yeah, I agree.

Is it a hassle to update QR Manager as you described?

Comment (CaseyODonnell92) [reply src/main/java/io/electrum/qr/api/model/CreateQrCodeMessage.java:44]: I don't know exactly what it ought to look like just yet... Might just be in the form of multiple owner accounts? (i.e. you can pay me at one of these accounts after scanning my QR). At any rate, we know it makes sense here, so we will make it work in QR Manager one way or another.

File: src/test/java/io/electrum/qr/api/model/NewModelTest.java [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+                                                      .city("Cape Town")
+                                                      .region("WC")
+                                                      .country("ZA"))))
+                  .tranId("V337HvR6TLWy+URFFLWE4g")
+                  .qrCode("2058443462");
+
+      errorDetail =
+            new ErrorDetail().id("c01cafdb-6ecc-4612-8acf-2d8dfd04a624")
+                  .errorMessage("Issuer Unavailable")
+                  .errorType(ErrorDetail.ErrorType.UPSTREAM_UNAVAILABLE);
```
Comment (CaseyODonnell92) [inline src/test/java/io/electrum/qr/api/model/NewModelTest.java:104]:
```suggestion
            new ErrorDetail().id("c01cafdb-6ecc-4612-8acf-2d8dfd04a624")
                  .originalId("b46f7ddb-c73c-496c-8849-2a3396aac861")
                      .errorType(ErrorDetail.ErrorType.UPSTREAM_UNAVAILABLE)
                      .errorMessage("Issuer Unavailable")
                      .detailMessage("This is a long explanation of why this message failed.")
                      .providerErrorCode("ProvideError")
                      .providerErrorMsg("Provider error message: we experienced a server error")
                      .providerRef("abcdef1234")
                      .tranId("V337HvR6TLWy+URFFLWE4g");
```
If I'm not mistaken, the goal of these tests is to detect any problems with changes to any of the model definitions that are unintentionally breaking. I think that we accomplish that better if we cover all of the top-level fields of every model definition.

File: src/test/java/io/electrum/qr/api/model/NewModelTest.java [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+                        Arrays.asList(
+                              new ThirdPartyIdentifier().institutionId("632940").transactionIdentifier("PP0715174834"),
+                              new ThirdPartyIdentifier().institutionId("0027").transactionIdentifier("2002441863"),
+                              new ThirdPartyIdentifier().institutionId("63294523").transactionIdentifier("601431"),
+                              new ThirdPartyIdentifier().institutionId("0027").transactionIdentifier("2002441863")))
+                  .rrn("PC0715174834");
+
+      paymentRequest =
+            new PaymentRequest().tranId("V337HvR6TLWy+URFFLWE4g")
+                  .amounts(new Amounts().requestAmount(new LedgerAmount().amount(12189L).currency("710")))
```
Comment (CaseyODonnell92) [inline src/test/java/io/electrum/qr/api/model/NewModelTest.java:122]:
```suggestion
            new PaymentRequest().tranId("V337HvR6TLWy+URFFLWE4g")
                  .partner(new Institution().id("0027").name("Masterpass"))
                  .partnerPaymentToken("6731210090")
                  .amounts(new Amounts().requestAmount(new LedgerAmount().amount(12189L).currency("710")))
```

File: src/test/java/io/electrum/qr/api/model/NewModelTest.java [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+                        Arrays.asList(
+                              new ThirdPartyIdentifier().institutionId("632940").transactionIdentifier("PP0715180220"),
+                              new ThirdPartyIdentifier().institutionId("0027").transactionIdentifier("2002460628"),
+                              new ThirdPartyIdentifier().institutionId("63294523").transactionIdentifier("739986"),
+                              new ThirdPartyIdentifier().institutionId("0027").transactionIdentifier("2002460628")))
+                  .time(DateTime.parse("2021-07-15T16:02:20.913Z"))
+                  .rrn("833057409");
+
+      paymentReversal =
+            new PaymentReversal().tranId("LKdTE6nCQlyrqmQe5ZhczQ")
```
Comment (CaseyODonnell92) [inline src/test/java/io/electrum/qr/api/model/NewModelTest.java:187]:
```suggestion
            new PaymentReversal().tranId("LKdTE6nCQlyrqmQe5ZhczQ")
                              .partner(new Institution().id("0027").name("Masterpass"))

```

File: src/test/java/io/electrum/qr/api/model/NewModelTest.java [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+      scanNotification =
+            new ScanNotification().settlementEntity(new Institution().id("63290").name("The Entity"))
+                  .tranId("PP0000000000")
+                  .time(DateTime.parse("2021-07-15T15:49:01.452Z"))
+                  .id("e104f4b4-2ef8-47bc-ab8a-1c6befff5d99")
+                  .partner(new Institution().id("10111").name("Police"))
+                  .partnerPaymentToken("1234567890")
+                  .receiver(new Institution().id("51290").name("The Five O"))
+                  .amounts(new Amounts().requestAmount(new LedgerAmount().amount(100L).currency("710")))
+                  .thirdPartyIdentifiers(Arrays.asList());
```
Comment (CaseyODonnell92) [inline src/test/java/io/electrum/qr/api/model/NewModelTest.java:211]: If we leave this empty and someone were to change the type of this field from `List<ThirdPartyIdentifier>` to any other kind of list, this test wouldn't pick that up. So we need at least one of those in there.

**Review by CaseyODonnell92 — CHANGES_REQUESTED** [review]
Please re-add the test class you removed. It's still valuable and if you have a look at Tim's comment, you'll see he still wanted it to hang around. Tests look good! Just gotta add some missing fields to make these tests really sensitive to unintended changes.

File: src/main/java/io/electrum/qr/api/model/CreateQrCodeRequest.java [hunk 17]
```diff
@@ -82,8 +82,9 @@ public CreateQrCodeRequest qrProperties(QrProperties qrProperties) {
 
    @Override
    public String toString() {
-      StringBuilder sb =
-            new StringBuilder("class CreateQrCodeRequest {}").append(System.lineSeparator()).append(super.toString());
+      StringBuilder sb = new StringBuilder("class CreateQrCodeRequest {").append(System.lineSeparator());
```
Comment (CaseyODonnell92) [inline src/main/java/io/electrum/qr/api/model/CreateQrCodeRequest.java:85]: All of the other `toString` have this pattern: print the child class fields within one set of curly braces, and then print the result of `super.toString()` after that. Is there a reason you changed this specific class's implementation? If not, no need for changes.

File: src/test/java/io/electrum/qr/api/model/ModelsTest.java [hunk 18]
```diff
@@ -1,4 +1,4 @@
-package io.electrum.qr.api;
+package io.electrum.qr.api.model;
```
Comment (CaseyODonnell92) [inline src/test/java/io/electrum/qr/api/model/ModelsTest.java:1]: I think you put this file back in a different package. Probably not a problem?

**Review by CaseyODonnell92 — CHANGES_REQUESTED** [review]
Just a little worried about the toString/equals changes. Otherwise we're done!
